### PR TITLE
Java: Convert SQL sinks to CSV format

### PR DESCRIPTION
**Pull request by joefarebrother**
Additionally, some missing sinks for android SQLite are added

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/SpringJdbc.qll [hunk 1]
```diff
@@ -3,33 +3,27 @@
...
-    this.getDeclaringType() instanceof JdbcTemplate and
-    this.hasName("batchUpdate") and
-    this.getParameterType(0).(Array).getComponentType() instanceof TypeString and
-    this.getParameter(0).isVarargs()
+private class SqlSinkCsv extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;spec;kind"
+        "org.springframework.jdbc.core;JdbcTemplate;false;batchUpdate;;;Argument[0];sql",
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/SpringJdbc.qll:18]: Missing the varargs case, which is now possible to express:
```suggestion
        "org.springframework.jdbc.core;JdbcTemplate;false;batchUpdate;(String[]);;ArrayElement of Argument[0];sql",
        "org.springframework.jdbc.core;JdbcTemplate;false;batchUpdate;;;Argument[0];sql",
```

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 2]
```diff
@@ -23,209 +24,77 @@ class TypeDatabaseUtils extends Class {
...
+        // query(String table, String[] columns, String selection, String[] selectionArgs, String groupBy, String having, String orderBy, String limit)
+        // query(String table, String[] columns, String selection, String[] selectionArgs, String groupBy, String having, String orderBy)
+        // queryWithFactory(SQLiteDatabase.CursorFactory cursorFactory, boolean distinct, String table, String[] columns, String selection, String[] selectionArgs, String groupBy, String having, String orderBy, String limit, CancellationSignal cancellationSignal)
+        // queryWithFactory(SQLiteDatabase.CursorFactory cursorFactory, boolean distinct, String table, String[] columns, String selection, String[] selectionArgs, String groupBy, String having, String orderBy, String limit)
+        // Each String / String[] arg except for selectionArgs is a sink
+        "android.database.sqlite;SQLiteDatabase;false;query;(String,String[],String,String[],String,String,String,String);;Argument[0..2];sql",
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:43]: Argument 1 looks like it ought to include `ArrayElement of`?

File: java/ql/src/semmle/code/java/frameworks/android/SQLite.qll [hunk 3]
```diff
@@ -23,209 +24,77 @@ class TypeDatabaseUtils extends Class {
...
+        "android.database.sqlite;SQLiteDatabase;false;updateWithOnConflict;(String,ContentValues,String,String[],int);;Argument[2];sql",
+        "android.database;DatabaseUtils;false;longForQuery;(SQLiteDatabase,String,String[]);;Argument[1];sql",
+        "android.database;DatabaseUtils;false;stringForQuery;(SQLiteDatabase,String,String[]);;Argument[1];sql",
+        "android.database;DatabaseUtils;false;blobFileDescriptorForQuery;(SQLiteDatabase,String,String[]);;Argument[1];sql",
+        "android.database;DatabaseUtils;false;createDbFromSqlStatements;(Context,String,int,String);;Argument[3];sql",
+        "android.database;DatabaseUtils;false;queryNumEntries;(SQLiteDatabase,String);;Argument[1];sql",
+        "android.database;DatabaseUtils;false;queryNumEntries;(SQLiteDatabase,String,String);;Argument[1..2];sql",
+        "android.database;DatabaseUtils;false;queryNumEntries;(SQLiteDatabase,String,String,String[]);;Argument[1..2];sql",
+        "android.database.sqlite;SQLiteQueryBuilder;false;delete;(SQLiteDatabase,String,String[]);;Argument[-1];sql",
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/android/SQLite.qll:71]: The `SQLiteQueryBuilder` method looks like it included overrides. Should we s/false/true/ here?

File: java/ql/test/library-tests/frameworks/android/taint-database/FlowSteps.java [hunk 4]
```diff
@@ -27,155 +27,160 @@
...
-		String[] newValues = taint();
+		String[] originalValues = {taint()}; // $hasTaintFlowStep
+		String[] newValues = {taint()}; // $hasTaintFlowStep
 		return DatabaseUtils.appendSelectionArgs(originalValues, newValues);
 	}
 
 	public static String concatenateWhere() {
-		String a = taint();
-		String b = taint();
+		String a = taint(); // $hasTaintFlowStep
```
Comment (aschackmull) [inline java/ql/test/library-tests/frameworks/android/taint-database/FlowSteps.java:36]: I think this test becomes a bit easier to read if we rename the label to something like "taintReachesReturn" - that way you don't need to look into the flow configuration to figure out what's being tested.

File: java/ql/test/library-tests/frameworks/android/taint-database/Sinks.java [hunk 5]
```diff
@@ -213,147 +213,147 @@ public static void query(MyContentResolver target) {
...
-		String sqlStatements = taint();
+		String sqlStatements = taint(); // $hasTaintFlowSink
 		DatabaseUtils.createDbFromSqlStatements(context, dbName, dbVersion, sqlStatements);
 	}
 
 	public static void blobFileDescriptorForQuery() {
 		SQLiteDatabase db = taint();
-		String query = taint();
-		String[] selectionArgs = taint();
+		String query = taint(); // $hasTaintFlowSink
```
Comment (aschackmull) [inline java/ql/test/library-tests/frameworks/android/taint-database/Sinks.java:341]: Similarly here, I think the label could be a bit clearer. E.g. "taintReachesQueryInjectionSink".

**Review by aschackmull — APPROVED** [review]

**Comment by joefarebrother** [comment]
Do we want to wait for sinks involving `ArrayElement of` to work properly before merging this? (for now the corresponding test cases are marked as missing)

**Comment by aschackmull** [comment]
With the recently merged https://github.com/github/codeql/pull/6107 we now have support for reading an ArrayContent at a sink, but giving a precise sink specification like `ArrayElement of Argument[0]` still doesn't work - it has to be just `Argument[0]` (which will then permit `ArrayElement of` if the type of the argument is an array-type). So let's remove `ArrayElement of` from the sink spec and get this merged.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**